tmuxPlugins.tmux-window-name: init at unstable

ndom91 opened the pull request:
## Description of changes

- Adds [tmux-window-name](https://github.com/ofirgall/tmux-window-name/tree/master) tmux plugin to `tmuxPlugins`.
- This plugin automatically renames windows based on the active windows active process and/or current working directory.
- This plugin requires interacting with the python `libtmux`, whose path is hardcoded into the shipped python scripts. So I did a few `sed` replacements in the `postInstall` step to patch for NixOS.
	- First, the entry file runs `python3 -m pip list` to check whether `libtmux` is available. This is handled by the wrapped script and its python env
	- Then, since its point is to check the running process and modify tmux window names to the respective applications, it required changes of the path it checks for the binaries, for example, to `$HOME/.nix-profile/bin/*` in a few places. 
	- It does a good bit of regex replacement to get the active program, active path, etc. so those needed patching as well
	
	
<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

ndom91 commented:
![image](https://github.com/NixOS/nixpkgs/assets/7415984/9fc76968-745b-4919-8c30-cdbeb10756e9)


The `nixpkgs-review` cmd is failing on a `merge` step. This is my first PR to nixpkgs, so not entirely sure what's going on here or if that's strictly necessary to run. 

Everything works locally and other CI steps look good :crossed_fingers:

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/trying-to-setup-tmux-plugin-that-requires-python3-and-libtmux-derivation-help/41028/6

ndom91 commented:
Ah so it looks like some of the changes were cached / not reapplied correctly. After a reboot it looks like the current bottom 2 `sed` commands in the `postIntsall` step (for `substitute_sets` and `dir_programs`) aren't workign as expected. 

That means the following two settings are still required:

```tmux
set -g @tmux_window_name_substitute_sets "[('/home/${config.home.username}/.nix-profile/bin/(.+) --.*', '\\g<1>')]"
set -g @tmux_window_name_dir_programs "['nvim', 'vim', 'vi', 'git', '/home/${config.home.username}/.nix-profile/bin/nvim']"
```

Should I add merge these into the user's `programs.tmux.extraConfig`? That's done with `lib.mkMerge`, right?

jeantil commented:
Hello @ndom91 
did you ever figure out how to finish this PR ? I am new to nix, trying to port my terminal setup and very sad not to find this tmux plugin :D

ndom91 commented:
> Hello @ndom91 did you ever figure out how to finish this PR ? I am new to nix, trying to port my terminal setup and very sad not to find this tmux plugin :D

Hey @jeantil yeah so I got it up and running in my system basically just with the above mentioned setup. Here are the finals:

- [tmux-window-name](https://github.com/ndom91/nixos-config/blob/main/modules/home-manager/tmux.nix#L7-L34)
- And the using it below in [that same file](https://github.com/ndom91/nixos-config/blob/main/modules/home-manager/tmux.nix#L46-L54).

FliegendeWurst commented on pkgs/misc/tmux-plugins/default.nix:
```diff
@@ -819,6 +819,46 @@ in rec {
+      description = "A plugin to name your tmux windows smartly, like IDE's";
+      license = licenses.mit;
+      platforms = platforms.unix;
+      maintainers = with maintainers; [ ndom91 ];
```
You also need to add yourself to maintainer-list.nix

ndom91 replied:
This went under, found it again thanks to the recent comment. 

I've added myself to the maintainer-list.nix in this branch and applied your other suggestion.

FliegendeWurst commented on pkgs/misc/tmux-plugins/default.nix:
```diff
@@ -819,6 +819,46 @@ in rec {
+    '';
+    meta = with lib; {
+      homepage = "https://github.com/ofirgall/tmux-window-name";
+      description = "A plugin to name your tmux windows smartly, like IDE's";
```
```suggestion
      description = "Tmux plugin to name your windows smartly, like IDE's";
```

phucisstupid commented:
any update

ndom91 commented:
Regarding the CI failures due to the merge commits not following conventional commits - is that okay? Or should I go thorugh and force push to rewrite the commit message for those?

FliegendeWurst commented:
It would be good to rebase the branch and fix the commit messages, yes. I think the PR cannot be merged otherwise.

ndom91 committed:
tmux-plugins: add tmux-window-name

ndom91 committed:
tmux-plugins.tmux-window-name: update description

Co-authored-by: Arne Keller <arne.keller@posteo.de>

ndom91 committed:
maintainers: add ndom91

ndom91 committed:
tmux-plugins: reformat

ndom91 committed:
tmux-plugins.tmux-window-name: resolve home at runtime

Replace the prior config.home.username reference (a home-manager
attribute unavailable in nixpkgs eval context) with a sed substitution
that injects os.path.expanduser("~") so the python script resolves the
user's home directory at runtime. The 'os' module is already imported
by upstream rename_session_windows.py.

ndom91 commented:
> It would be good to rebase the branch and fix the commit messages, yes. I think the PR cannot be merged otherwise.

Okay great, I've done that. Let's see how CI goes now 🤞
